sdk/eventhub/azure-eventhub/tests/livetest/synctests/test_reconnect.py

```diff
@@ -164,7 +164,8 @@ def on_event_received(event):
                 consumer._handler.do_work()
                 assert consumer._handler._connection._state == uamqp.c_uamqp.ConnectionState.DISCARDING
             else:
-                consumer._handler.do_work()
+                with pytest.raises(error.AMQPConnectionError):
```

Comment:
https://github.com/Azure/azure-sdk-for-python/pull/30734/files#r1273745456

Reply:
OOB: The error was previously raised [[here in connection.listen](https://github.com/Azure/azure-sdk-for-python/blob/157226ef9713bb2ca248ea7ddb5284496ac0338b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_connection.py#L757)] and now it's raised [[here in the connection process_outgoing_frame](https://github.com/Azure/azure-sdk-for-python/blob/157226ef9713bb2ca248ea7ddb5284496ac0338b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_connection.py#L677)] since the the flow is called before listen can be.

The previous logs looked like this when do_work was called, with session end not being sent in the do_work.
```
before do work
CBS status check: state == <CbsAuthState.OK: 0>, expired == False, refresh required == False
No frame received for the idle timeout. Closing connection.
-> CloseFrame(error=AMQPError(condition=<ErrorCondition.ConnectionCloseForced: b'amqp:connection:forced'>, description='No frame received for the idle timeout.', info=None))
Connection state changed: <ConnectionState.OPENED: 9> -> <ConnectionState.DISCARDING: 12>
Connection state changed: <ConnectionState.DISCARDING: 12> -> <ConnectionState.END: 13>
Session state changed: <SessionState.MAPPED: 3> -> <SessionState.DISCARDING: 6>
Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
Management link sender state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
CBS error occurred.
Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
Management link receiver state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
after do work
Session state changed: <SessionState.DISCARDING: 6> -> <SessionState.END_SENT: 4>
```

Whereas with this fix, session end is sent, then error is raised within do_work.
```
before do work
CBS status check: state == <CbsAuthState.OK: 0>, expired == False, refresh required == False
-> FlowFrame(next_incoming_id=2, incoming_window=65535, next_outgoing_id=1, outgoing_window=65535, handle=3, delivery_count=0, link_credit=300, available=None, drain=None, echo=None, properties=None)     
No frame received for the idle timeout. Closing connection.
-> CloseFrame(error=AMQPError(condition=<ErrorCondition.ConnectionCloseForced: b'amqp:connection:forced'>, description='No frame received for the idle timeout.', info=None))
Connection state changed: <ConnectionState.OPENED: 9> -> <ConnectionState.DISCARDING: 12>
Connection state changed: <ConnectionState.DISCARDING: 12> -> <ConnectionState.END: 13>
Session state changed: <SessionState.MAPPED: 3> -> <SessionState.DISCARDING: 6>
Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
Management link sender state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
CBS error occurred.
Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
Management link receiver state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>
Session state changed: <SessionState.DISCARDING: 6> -> <SessionState.END_SENT: 4>
```

This new behavior with session end sent within do_work seems correct.